### PR TITLE
[BACKLOG-9118] Hide legends in CCC charts based on a maximum number of items

### PR DESCRIPTION
**Pull request by davidmsantos90**
Also removed some build artifacts that were committed to github.

@nantunes @graimundo @dcleao @marcovala @carlosrusso please review

### CODE REVIEW

File: doc/model/panels/legend.xml [hunk 1]
```diff
@@ -95,6 +95,24 @@
             	</p>
             </c:documentation>
         </c:property>
+
+        <c:property name="itemCountMax" type="number" category="Layout">
+            <c:documentation>
+                The maximum number of legend items that can feet in the legend panel.
```
Comment (jheck123) [inline doc/model/panels/legend.xml]: The maximum number of legend items that can fit in the legend panel.

File: doc/model/panels/legend.xml [hunk 2]
```diff
@@ -95,6 +95,24 @@
...
+
+                See <c:link to="pvc.options.varia.LegendOverflow" /> for information about
+                the different behaviours of the legend panel when this maximum is exceeded.
```
Comment (jheck123) [inline doc/model/panels/legend.xml]: the different behaviors of the legend panel when this maximum is exceeded.

File: doc/model/panels/legend.xml [hunk 3]
```diff
@@ -95,6 +95,24 @@
...
+            </c:documentation>
+        </c:property>
+
+        <c:property name="legendOverflow" type="string pvc.options.varia.LegendOverflow" category="Layout">
+            <c:documentation>
+                The behaviour of the legend panel when an overflow is detected.
```
Comment (jheck123) [inline doc/model/panels/legend.xml]: The behavior of the legend panel when an overflow is detected.

File: doc/model/panels/legend.xml [hunk 4]
```diff
@@ -95,6 +95,24 @@
...
+
+                See <c:link to="pvc.options.varia.LegendOverflow" /> for information about
+                the different behaviours of the legend panel when an overflow is detected.
```
Comment (jheck123) [inline doc/model/panels/legend.xml]: the different behaviors of the legend panel when an overflow is detected.

File: doc/model/varia/atoms/legendOverflow.xml [hunk 5]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<c:model
+        xmlns:c="urn:webdetails/com/2012"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="urn:webdetails/com/2012 ../../../schema/com_2012.xsd"
+        xmlns="http://www.w3.org/1999/xhtml">
+
+    <c:atomType name="LegendOverflow" space="pvc.options.varia" base="string">
+        <c:documentation>
+            The possible strategies for when the number of legend items exceeds
```
Comment (jheck123) [inline doc/model/varia/atoms/legendOverflow.xml]: The possible strategies if the number of legend items exceeds

File: package-res/ccc/core/base/panel/legend/scene/legend-root-scene.js [hunk 6]
```diff
@@ -104,6 +108,10 @@ def
         if(!$w || $w < 0) $w = contentSize[a_width ];
         if(!$h || $h < 0) $h = contentSize[a_height];
 
+        // If the number of items exceed the maximum count,
```
Comment (jheck123) [inline package-res/ccc/core/base/panel/legend/scene/legend-root-scene.js]: If the number of items exceeds the maximum count,

File: package-res/ccc/core/base/panel/legend/scene/legend-root-scene.js [hunk 7]
```diff
@@ -104,6 +108,10 @@ def
...
+        // the sections width or height don't have enough space to render
```
Comment (jheck123) [inline package-res/ccc/core/base/panel/legend/scene/legend-root-scene.js]: the section's width or height don't have enough space to render

File: doc/model/varia/atoms/legendOverflow.xml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            the <c:link to="pvc.options.panels.ItemCountMax" /> or overflows the
+            legend panel dimensions.
+        </c:documentation>
+
+        <c:atom name="Collapse" value="'collapse'">
+            <c:documentation>
+                When overflow is detected the whole legend panel
```
Comment (jheck123) [inline doc/model/varia/atoms/legendOverflow.xml]: When overflow is detected, the whole legend panel

File: doc/model/varia/atoms/legendOverflow.xml [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+                is collapsed and the legend is not displayed.
+            </c:documentation>
+        </c:atom>
+
+        <c:atom name="Clip" value="'clip'">
+            <c:documentation>
+                Only the first legend items that entirely fit
```
Comment (dcleao) [inline doc/model/varia/atoms/legendOverflow.xml]: Do they need to entirely fit? I think they can be partially inside.

File: doc/model/panels/legend.xml (hunk 2 above)
Comment (dcleao) [inline doc/model/panels/legend.xml]: Should add a note like "A positive number.".

File: doc/model/panels/legend.xml [hunk 10]
```diff
@@ -95,6 +95,24 @@
...
+        <c:property name="itemCountMax" type="number" category="Layout">
+            <c:documentation>
+                The maximum number of legend items that can feet in the legend panel.
+
+                See <c:link to="pvc.options.varia.LegendOverflow" /> for information about
+                the different behaviours of the legend panel when this maximum is exceeded.
+            </c:documentation>
+        </c:property>
+
+        <c:property name="legendOverflow" type="string pvc.options.varia.LegendOverflow" category="Layout">
```
Comment (dcleao) [inline doc/model/panels/legend.xml]: Does this not have a default value?

File: package-res/ccc/core/base/panel/legend/scene/legend-root-scene.js [hunk 11]
```diff
@@ -181,6 +191,16 @@ def
             // New section
             if(!isLast) section = new pvc.visual.legend.LegendItemSceneSection(sections.length);
         }
+
+        function isOverflow(itemCountMax) {
+            var desiredWidth = desiredClientSize.width || Infinity;
+            var desiredHeight = desiredClientSize.height || Infinity;
+
+            return (totalItemCount > itemCountMax) ||
```
Comment (dcleao) [inline package-res/ccc/core/base/panel/legend/scene/legend-root-scene.js]: Please separate the item count test and do it before any other calculations are performed.

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
+define([
+  'ccc/pvc',
+  'ccc/def',
+  'test/utils',
+  'test/data-1'
+], function(pvc, def, utils, datas) {
+
+  describe("legend-panel -", function() {
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: Please remove all end of line punctuation from describe/it. The new karma reporter does not need this.

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        it("'legendItemCountMax' isn't taken into account in 'Clip Mode'", function() {
+          var chartOptions = createOptions({legendOverflow: 'clip', legendItemCountMax: 2});
+          var actual = createChart(chartOptions);
+
+          expect(chartOptions.legendItemCountMax < actual.legendItems).toBe(true);
+          expect(actual.isLegendVisible).toBe(true);
+
+        });
+
+        it("'legendSize' will set the size of the legend panel and hide the items if they don't feet it", function() {
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: feet -> fit

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+          var chartOptions = createOptions({legendOverflow: 'clip', legendSize: size});
+          var actual = createChart(chartOptions);
+
+          expect(actual.legendItems).toBe(3);
+          expect(actual.isLegendVisible).toBe(true);
+        });
+
+        it("'legendSizeMax' will set the size of the legend panel and hide the items if they don't feet it", function() {
+          var chartOptions = createOptions({legendOverflow: 'clip', legendSize: size});
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: feet -> fit

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    describe("legendOverflow -", function() {
+      var size = {width: 50, height: 50};
+
+      describe("- Clip Mode", function() {
+
+        it("'legendItemCountMax' isn't taken into account in 'Clip Mode'", function() {
+          var chartOptions = createOptions({legendOverflow: 'clip', legendItemCountMax: 2});
+          var actual = createChart(chartOptions);
+
+          expect(chartOptions.legendItemCountMax < actual.legendItems).toBe(true);
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: Please invert the order legendItems > legendItemCountMax and use the karma toBeGreaterThan operator.

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+      });
+
+    });
+    
+  });
+
+  function createChart(options) {
+    var chart = new pvc.BarChart(options);
+    var dataSpec = datas['relational, series=city|category=date|value=qty, square form'];
+    dataSpec[1].crosstabMode = false;
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: The data spec is shared by several unit tests and so cannot be changed this way.
Please create a new data spec if you need to.

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+
+    chart.allowNoData = true;
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: Any reason why you need allowNoData in this case? If so please NOTE which.

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    chart.setData.apply(chart, dataSpec);
+
+    chart._create(options);
+    chart.render();
+
+    chart.legendPanel.layout({width: options.width, height: options.height}, {force: 1});
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: Why call chart.render and then repeat the legend layout? If there is any special reason, please NOTE which.

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+
+    return {
+      chart: chart,
+      legendItems: chart.legendPanel.data.getSeries().length,
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: --> legendItemCount ?

Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: data.getSeries is @deprecated AFAIK. please see how the legend scenes are built from data and use a similar way to access the correct data.

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 20]
```diff
@@ -0,0 +1,111 @@
+define([
+  'ccc/pvc',
+  'ccc/def',
+  'test/utils',
+  'test/data-1'
+], function(pvc, def, utils, datas) {
+
+  describe("legend-panel -", function() {
+
+    describe("legendOverflow -", function() {
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: please try to phrase the it("") titles as "should ..." .

File: test/ccc/core/base/panel/legend/legend-panel-spec.js (hunk 13 above)
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: What does the title of this unit test have to do with what is being tested?

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+
+        it("'legendSize' will set the size of the legend panel and hide the items if they don't feet it", function() {
+          var chartOptions = createOptions({legendOverflow: 'clip', legendSize: size});
+          var actual = createChart(chartOptions);
+
+          expect(actual.legendItems).toBe(3);
+          expect(actual.isLegendVisible).toBe(true);
+        });
+
+        it("'legendSizeMax' will set the size of the legend panel and hide the items if they don't feet it", function() {
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: Idem... The test title includes two cases, but only one (the least expected/relevant?) is tested.

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+          var actual = createChart(chartOptions);
+
+          expect(actual.legendItems).toBe(3);
+          expect(actual.isLegendVisible).toBe(true);
+        });
+
+      });
+
+      describe("- Collapse Mode", function() {
+        it("Should hide the legend panel when the legend items exceed the 'legendItemCountMax' property", function() {
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: Please use "should ..."

**Comment by davidmsantos90** [comment]
@dcleao Just finished going through your reviews.
@graimundo @marcovala @nelsonsousa please review.

File: doc/model/panels/legend.xml [hunk 23]
```diff
@@ -95,6 +95,24 @@
             	</p>
             </c:documentation>
         </c:property>
+
+        <c:property name="itemCountMax" type="number" category="Layout">
+            <c:documentation>
+                The maximum number of legend items that can fit in the legend panel.
```
Comment (dcleao) [inline doc/model/panels/legend.xml]: Please mention that the number should be positive.

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    return itemsTotalSize;
+  }
+
+  function _legendPanelSize(layout) {
+    var panelSize = { width: layout.clientSize.width, height: layout.clientSize.height };
+
+    if(panelSize.width > 0) {
+      panelSize.width += layout.realMargins.width + layout.realPaddings.width;
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: Please do not use realMargins and realPaddings (they are being removed from another working branch). Use layout.margins and layout.paddings instead. A new version will also already come with layoutInfo.size... but that's not yet usable...

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 25]
```diff
@@ -0,0 +1,184 @@
...
+      panelSize.height += layout.realMargins.height + layout.realPaddings.height;
+    }
+
+    return panelSize;
+  }
+
+  function _legendItemCount(rootScene) {
+    var count = 0;
+    rootScene.childNodes.forEach(function(scene) {
+      scene.childNodes.forEach(function () {
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: `count += scene.childNodes.length;`

File: test/ccc/core/base/panel/legend/legend-panel-spec.js [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+    if('legendSizeMax' in options) chartOptions.legendSizeMax = options.legendSizeMax;
+
+    return chartOptions;
+  }
+
+  function _itemsTotalSize(legendPanel) {
+    var itemsTotalSize = { width: 0, height: 0 };
+    if(!legendPanel.pvLegendPanel) return itemsTotalSize;
+
+    var sections = legendPanel.pvLegendPanel.data().vars.sections;
```
Comment (dcleao) [inline test/ccc/core/base/panel/legend/legend-panel-spec.js]: `var sections = legendPanel._rootScene.vars.sections;`

**Comment by dcleao** [comment]
👍

**Comment by davidmsantos90** [comment]
@dcleao just checked your final review comments. please review

File: doc/model/panels/legend.xml [hunk 27]
```diff
@@ -95,6 +95,24 @@
...
+        <c:property name="itemCountMax" type="number" category="Layout">
+            <c:documentation>
+                The maximum positive number of legend items that can fit in the legend panel.
+
+                See <c:link to="pvc.options.varia.LegendOverflow" /> for information about
+                the different behaviors of the legend panel when this maximum is exceeded.
+            </c:documentation>
+        </c:property>
+
+        <c:property name="legendOverflow" type="string pvc.options.varia.LegendOverflow" default="clip" category="Layout">
```
Comment (dcleao) [inline doc/model/panels/legend.xml]: `default="'clip'"`

**Comment by dcleao** [comment]
👍

**Comment by davidmsantos90** [comment]
@dcleao 'legendOverflow' default corrected. please review

**Pull request merged**